File: pipelines/path/temp/01.PIP-release/PIPv1.1.0rc0517/bin/4.Stat/Stat.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__author__ = 'sujiawei'
__email__ = 'su_jiawei163@163.com'
__date__ = '2022-03-31'
__version__ = 'V1.3'

import os
import sys
import click
import warnings
import pandas as pd
from pathlib import Path
from collections import defaultdict
from multiprocessing import Process

Bin = Path(__file__).resolve().parent
sys.path.append(os.path.join(Bin, '../../conf'))
import configure as cfg
from configure import checkFile, createPath, printInfo

warnings.filterwarnings('ignore')

# scripts
QC            = Bin / 'QC.stat.pl'
Normalize     = Bin / 'Normalize.py'
CommonSpecies = Bin / 'CommonSpecies.pl'
MergeResults  = Bin / 'mergeResults.py'
FillCoverage  = Bin / 'fillCoverage.py'
SplitQC       = Bin / 'split_QC_Stat.py'


def check_dir(path):
    if not Path(path).exists():
        os.system(f'mkdir -p {str(path)}')


def cp(src, tgt):
    os.system(f'cp -fr {str(src)} {str(tgt)} &')


def execute(cmd):
    os.system(cmd)


@click.command(no_args_is_help=True)
@click.option('-s', '--samples', help='samples list, Format: \"sample DNA|RNA read1.fq [read2.fq]\"\n'
                                      'There are 4 columns separated with Tab in the file, the 1st is \n'
                                      'sample name, the 2nd is sample type (DNA or RNA), the 3rd is the \n'
                                      'fastq file of reads (read1 for PE), the 4th is the fastq file of \n'
                                      'read2 (only for PE).')
@click.option('-a', '--analysis', help='The analytic steps you done. The different steps separated by \'-\', \n'
                                       'the arguments for every step follow the \':\' and separated by \',\'. \n'
                                       'The selectable steps as bollow: default:\n'
                                       '[Merge-QC:Fastp,RemoveHost,rRNAFilter-Annotation:Taxonomy,ResistanceGene,VFDB]')
@click.option('-w', '--workdir', help='The path including 00.Rawdata,01.QC,.... [./]')
def main(samples, analysis, workdir):
    workdir = Path(workdir)

    result_dir = workdir / 'plugin_out/Result'
    check_dir(result_dir)
    qc_dir  = workdir / '01.QC'
    tax_dir = workdir / '02.Annotation/01.Taxonomy'
    arg_dir = workdir / '02.Annotation/02.ResistanceGene'
    vf_dir  = workdir / '02.Annotation/03.VFDB'

    db_ver = analysis.split('=')[1].split(',')[0]
    db_path = Path(cfg.KrakenDB) / db_ver

    os.system(f'{cfg.perl} {QC} {str(qc_dir)}/sample.list {str(result_dir)}/QC.Stat')
    df_qc = pd.read_csv(result_dir / 'QC.Stat', sep='\t', dtype=str)
    df_qc.to_excel(result_dir / 'QC.Stat.xlsx', index=False)
    os.system(f'{cfg.python} {SplitQC} -d {str(workdir)}')
    os.system(f'{cfg.python} {Normalize} --qc {str(result_dir)}/QC.Stat --anno {str(tax_dir)}/sample.list')
    os.system(f'{cfg.perl} {CommonSpecies} --list {str(tax_dir)}/sample.list --out {str(result_dir)}')
    os.chdir(str(result_dir))
    os.system(f'zip -r {str(result_dir)}/zhikong.zip Total_CommonSpecies* Pathogeny_CommonSpecies* QC.Stat*')

    df = pd.read_csv(samples, sep='\t', dtype=str, header=None)

    # basecaller_results
    basecall_dir = workdir / 'basecaller_results'
    check_dir(basecall_dir)
    for row in df.itertuples():
        slide_dir = Path(row[3]).parent
        cp(slide_dir / '*.summaryReport.html', basecall_dir)

    # create by2.run
    with open(workdir / 'by2.run', 'w') as wfp:
        wfp.write('\n'.join(['Result'] + df[0].tolist()))

    process_lst = []
    seq_cp_list = []
    kingdom = ['bacteria', 'fungi', 'LY', 'protozoa', 'viral']
    for row in df.itertuples():
        smp_dir = result_dir / row[1]
        check_dir(smp_dir)
        cmd1 = f'{cfg.python} {FillCoverage} --db {str(db_path)} --indir {str(workdir)} --name {row[1]} ' \
               f'--library {row[2]} --outdir {str(smp_dir)}'
        p = Process(target=execute, args=(cmd1,))
        p.start()
        process_lst.append(p)
        # copy
        orig_tax_dir = tax_dir / row[1] / row[2]
        for kd in kingdom:
            p = Process(target=cp, args=(orig_tax_dir / f'{kd}.xls', smp_dir / f'{kd}.xls'))
            p.start()
            process_lst.append(p)
            p = Process(target=cp, args=(orig_tax_dir / f'{kd}.xls', smp_dir / f'{kd}.pick.xls'))
            p.start()
            process_lst.append(p)
        # ARG
        p = Process(target=cp, args=(arg_dir / row[1] / row[2] / 'ARG.cov.pick.xls', smp_dir))
        p.start()
        process_lst.append(p)
        # VF
        p = Process(target=cp, args=(vf_dir / row[1] / row[2] / 'VF.xls', smp_dir))
        p.start()
        process_lst.append(p)
        # merge pathogen results
        cmd2 = f'{cfg.python} {MergeResults} --workdir {str(workdir)} --name {row[1]} --library {row[2]}'
        print(cmd2)
        p = Process(target=execute, args=(cmd2,))
        p.start()
        process_lst.append(p)
        # by2.sample
        p = Process(target=execute, args=(f'touch {str(workdir)}/by2.sample',))
        p.start()
        process_lst.append(p)
        # execute process
        for p in process_lst:
            p.join()
        process_lst.clear()
        # copy sequences
        for kd in kingdom:
            p = Process(target=cp, args=(tax_dir / row[1] / row[2] / f'{kd}.result.fq.gz',
                                         smp_dir / f'{row[1]}.{kd}.result.fq.gz'))
            p.start()
            seq_cp_list.append(p)
        # copy summary
        for file in ['Pathogeny_Detail', 'Total_Detail', 'Taxonomy_Summary']:
            p = Process(target=cp, args=(tax_dir / row[1] / row[2] / f'{file}.*', smp_dir))
            p.start()
            process_lst.append(p)
        for p in process_lst:
            p.join()
        process_lst.clear()

    for p in seq_cp_list:
        p.join()

    os.system(f'touch {str(workdir)}/by2.done')


if __name__ == '__main__':
    main()

File: pipelines/path/temp/01.PIP-release/PIPv1.0.0rc4/bin/4.Stat/Stat.pl
#!/usr/bin/env perl

#===============================================================================
#   FileName: Stat.pl
#   Author  : qiuwancen
#   E-Mail  : 972538446@qq.com
#   Version : 1.0
#   Date    : 2022-03-22
#   Description: Copy the result files to Result directory.
#===============================================================================

use warnings;
use strict;
use FindBin qw($Bin);
use Cwd qw(abs_path);
use File::Basename;
use Getopt::Long;
use Pod::Usage;
use lib "$Bin/../../lib/perl5";
use MyModule::MyLog qw(printInfo2ERR);
use MyModule::GlobalVar qw($TXT2EXCEL $PERL $PYTHON3 $KRAKENDB);

my($list, $analysis, $workdir, $result, $shellDir, $help);
GetOptions(
    "analysis=s" => \$analysis,
    "workdir=s"  => \$workdir,
    "result=s"   => \$result,
    "shellDir=s" => \$shellDir,
    "help|h"     => \$help,
);
pod2usage(-exitval => 1, -verbose => 1) if($help || @ARGV < 1);
($list, ) = @ARGV;
$list = abs_path $list;


#===============================================================================
#   Set the default values of options
#===============================================================================
$analysis ||= "Merge-QC:Fastp,RemoveHost,rRNAFilter-Annotation:Taxonomy,ResistanceGene,VFDB";
$workdir  ||= "./";
$shellDir ||= "./";
$workdir  = abs_path $workdir;
$result   =   "$workdir/plugin_out/Result";
`mkdir -p $result` unless(-d $result);
$result   = abs_path $result;
$shellDir = abs_path $shellDir;

#===============================================================================
#   Create the analytic directory
#===============================================================================
#`mkdir -p $shellDir` unless(-d $shellDir);
`mkdir -p $workdir/basecaller_results` unless(-d "$workdir/basecaller_results");

#===============================================================================
#   Scripts
#===============================================================================
my $QC            = "$Bin/QC.stat.pl";
my $Normalize     = "$Bin/Normalize.py";
my $CommonSpecies = "$Bin/CommonSpecies.pl";
my $mergeResults  = "$Bin/mergeResults.py";
my $fillCoverage  = "$Bin/fillCoverage.py";
my $plot          = "$Bin/Plot.py";
my $split_qc      = "$Bin/split_QC_Stat.py";

#===============================================================================
#   Global variables
#===============================================================================
## Directory
my %path = (
    "QC"                => "$workdir/01.QC",
    "Taxonomy"          => "$workdir/02.Annotation/01.Taxonomy",
    "ResistanceGene"    => "$workdir/02.Annotation/02.ResistanceGene",
    "VFDB"              => "$workdir/02.Annotation/03.VFDB",
);

#===============================================================================
#   Get the sample list
#===============================================================================
my %samp_list = ();
open IN,"$list" or die $!;
my $flag = 0;
while (<IN>) {
    chomp;
    my($name, $type, $fq) = (split /\t/,$_)[0, 1, 2];
	if($flag == 0){
		my $dir = dirname $fq;
		`cp $dir/*.summaryReport.html $workdir/basecaller_results`;
	}
    my $sample = "$name\_$type";
    $samp_list{$name}{$type} = $sample;
}
close IN;


#===============================================================================
#   Generate the scripts
#===============================================================================
`$PERL $QC $path{QC}/sample.list $result/QC.Stat`;
`$PERL $TXT2EXCEL $result/QC.Stat $result/QC.Stat.xlsx`;
`$PYTHON3 $split_qc -d $workdir`;
`$PYTHON3 $Normalize --qc $result/QC.Stat --anno $path{Taxonomy}/sample.list`;
`$PERL $CommonSpecies --list $path{Taxonomy}/sample.list --outdir $result`;
#`$PYTHON3 $plot --qc $result/QC.Stat --profile $result/Pathogeny_CommonSpecies.Normalize.txt --outdir $result`;
chdir("$result");
`zip -r $result/zhikong.zip Total_CommonSpecies* Pathogeny_CommonSpecies* QC.Stat*`;

for my $name (keys %samp_list) {
    my $sampDir = "$result/$name";
    `mkdir -p $sampDir` unless(-d $sampDir);
	`rm $workdir/by2.sample` if(-e "$workdir/by2.sample");

    # Copy the files of identification
    if ($analysis =~ /Taxonomy/) {
    	my @tmp = split(/=/, $analysis);
    	my $db  = (split(/,/, $tmp[1]))[0];
        my $origTaxDir = $path{Taxonomy};
        for my $type (keys %{$samp_list{$name}}) {
			`$PYTHON3 $fillCoverage --db $KRAKENDB/$db --indir $workdir --name $name --library $type --outdir $sampDir`;
			`cp -fr $origTaxDir/$name/$type/*.xls $sampDir`;
			`cp -fr $origTaxDir/$name/$type/Pathogeny_Detail.xlsx $sampDir`;
			`cp -fr $origTaxDir/$name/$type/Pathogeny_Detail.txt $sampDir`;
			`cp -fr $origTaxDir/$name/$type/Total_Detail.xlsx $sampDir`;
			`cp -fr $origTaxDir/$name/$type/Total_Detail.txt $sampDir`;
			`cp -fr $origTaxDir/$name/$type/Taxonomy_Summary.txt $sampDir`;
			`cp -fr $origTaxDir/$name/$type/bacteria.result.fq.gz $sampDir/`; ###temp
			`cp -fr $origTaxDir/$name/$type/fungi.result.fq.gz $sampDir/`; ###temp
			`cp -fr $origTaxDir/$name/$type/protozoa.result.fq.gz $sampDir/`; ###temp
			`cp -fr $origTaxDir/$name/$type/viral.result.fq.gz $sampDir/`; ###temp
			`cp -fr $origTaxDir/$name/$type/LY.result.fq.gz $sampDir/`; ###temp
			`cp -fr $sampDir/bacteria.xls $sampDir/bacteria.pick.xls`;
			`cp -fr $sampDir/fungi.xls $sampDir/fungi.pick.xls`;
			`cp -fr $sampDir/LY.xls $sampDir/LY.pick.xls`;
			`cp -fr $sampDir/protozoa.xls $sampDir/protozoa.pick.xls`;
			`cp -fr $sampDir/viral.xls $sampDir/viral.pick.xls`;
			`$PYTHON3 $mergeResults --workdir $workdir --name $name --library $type`;
        }
    }
    
    # Copy the files of VF identification
    if ($analysis =~ /VFDB/) {
        my $origVfDir = $path{"VFDB"};
        for my $type (keys %{$samp_list{$name}}) {
            `cp -fr $origVfDir/$name/$type/VF.xls $sampDir`;
        }
    }

	# Copy the files of ResistanceGene identification
	if ($analysis =~ /ResistanceGene/) {
		my $origResGeneDir = $path{"ResistanceGene"};
		for my $type (keys %{$samp_list{$name}}){
			`cp -fr $origResGeneDir/$name/$type/ARG.cov.pick.xls $sampDir`;
		}
	}
    
    # Copy the files of RNA vs DNA
#    if($analysis =~ /RNADNA/){
#        my $origVsDir = $path{"RNADNA"};
#        my $destVsDir = $destPath{"RNADNA"};
#        system("mkdir -p $destVsDir") unless (-e $destVsDir);
#        print SH "cp -fr $origVsDir/Result/$name/* $destVsDir\n";
#    }

	# Copy the files of SpeciesExtract
#	if($analysis =~ /SpeciesExtract/){
#		my $origExtractDir = $path{"SpeciesExtract"};
#		my $destExtractDir = $destPath{"SpeciesExtract"};
#		for my $type (keys %{$samp_list{$name}}){
#			my $destTypeDir = "$destExtractDir/$type";
#			if(-d "$origExtractDir/$name/$type"){
#				system("mkdir -p $destTypeDir") unless (-e $destTypeDir);
#				print SH "cp -fr $origExtractDir/$name/$type/* $destTypeDir\n";
#			}
#		}
#	}
}
#close SH;

###############################################################################################################

open RUN,">$workdir/by2.run" or die $!;
print RUN "Result\n";
for my $name (keys %samp_list) {
	print RUN "$name\n";
}
close RUN;

`sleep 30`;

`touch $workdir/by2.sample`;
`touch $workdir/by2.done`;

#===============================================================================
#   Execute the scripts
#===============================================================================
#printInfo2ERR("Start program $0");
#`sh $shellDir/upload.sh`;
#printInfo2ERR("Finish program $0");

__END__

=pod

=head1 NAME

upload.pl - The program for uploading result

=head1 VERSION

v1.0

=head1 SYNOPSIS

perl upload.pl <sample.list> [options]

=head1 ARGUMENTS

=over 8

=item B<sample.list> <file>

List of raw Reads.

Format: "sample DNA|RNA read1.fq [read2.fq]"

There are 4 columns separated with Tab in the file, the 1st is sample name, the
2nd is sample type (DNA or RNA), the 3rd is the fastq file of reads (read1 for PE),
the 4th is the fastq file of read2 (only for PE).

=back

=head1 OPTIONS

=over 8

=item B<--analysis> <Str>

The analytic steps you done. The different steps separated by '-', the
arguments for every step follow the ':' and separated by ','. The selectable
steps as bollow:
default: [Merge-QC:Fastp,RemoveHost,rRNAFilter-Annotation:Taxonomy,ResistanceGene,VFDB]

=item B<--workdir> <Path>

The path including 00.Rawdata,01.QC,.... [./]

=item B<--result> <Path>

Result path. [outdir/Result]

=item B<--shellDir> <path>

The directory to store the scripts generated by this program. [./]

=item B<--help|h>

Print this information.

=back

=cut
